convert-hf-to-gguf.py

```diff
@@ -2936,6 +2939,93 @@ def modify_tensors(self, data_torch: Tensor, name: str, bid: int | None) -> Iter
         return [(self.map_tensor_name(name), data_torch)]
 
 
+@Model.register("JAISLMHeadModel")
+class JaisModel(Model):
+    model_arch = gguf.MODEL_ARCH.JAIS
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        # SwigLU activation
+        assert self.hparams["activation_function"] == "swiglu"
+        # ALiBi position embedding
+        assert self.hparams["position_embedding_type"] == "alibi"
+
+        # Embeddings scale
+        self.embeddings_scale = 1.0
+        # note: For some JAIS flavors, output is tied to (same as) wte in original model
+        self.output_is_wte = False
+        if 'mup_embeddings_scale' in self.hparams:
+            self.output_is_wte = True   # Hack (?)
+            self.embeddings_scale = self.hparams['mup_embeddings_scale']
+        elif 'embeddings_scale' in self.hparams:
+            self.embeddings_scale = self.hparams['embeddings_scale']
+        else:
+            assert False
+
+        self.width_scale = 1.0
+        if 'mup_output_alpha' in self.hparams:
+            assert 'mup_width_scale' in self.hparams
+            self.width_scale = self.hparams['mup_output_alpha'] * self.hparams['mup_width_scale']
+        elif 'width_scale' in self.hparams:
+            self.width_scale = self.hparams['width_scale']
+        else:
+            assert False
+
+    def set_vocab(self):
+        self._set_vocab_gpt2()
+
+    def set_gguf_parameters(self):
+        self.gguf_writer.add_name(self.dir_model.name)
+        self.gguf_writer.add_block_count(self.hparams["n_layer"])
+        self.gguf_writer.add_context_length(self.hparams["n_positions"])
+        self.gguf_writer.add_embedding_length(self.hparams["n_embd"])
+        self.gguf_writer.add_feed_forward_length(self.hparams["n_inner"])
+        self.gguf_writer.add_head_count(self.hparams["n_head"])
+        self.gguf_writer.add_layer_norm_eps(self.hparams["layer_norm_epsilon"])
+        self.gguf_writer.add_file_type(self.ftype)
+
+        # Hack to populate self.tensor_names
+        all(self.get_tensors())
+        if 'transformer.relative_pe.slopes' not in self.tensor_names:
+            self.gguf_writer.add_max_alibi_bias(8.0)
+        # else set later
```

Comment:
Note that `self.tensor_names` is fully populated when even only the first tensor has been read from `self.get_tensors` (no need to read them `all`).

Alternatively (still hacky), the value could be set after `self.write_tensors` by overriding that method (a bit like you've done with `__init__`):

```python3
    def write_tensors(self, *args, **kwargs):
        super().write_tensors(*args, **kwargs)
        
        self.gguf_writer.add_max_alibi_bias(self._some_alibi_bias_var)
```
and that `self._some_alibi_bias_var` (with a better name) could be initially set to 8.0 in `__init__`, and updated in `modify_tensors`.

Reply:
`all` is definitely an overkill here. I was looking for a post-process step, and your solution of using write_tensors achieves exactly that!
I also do think that the value is a constant for this architecture, but I'm just having some harmless fun ;)